pkg/region/cache.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2021 Outreach Corporation. All Rights Reserved.
+
+// Description: This file contains the read through cache implementation for
+// this package.
+package region
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	// RegionCachePath is the $HOME/<RegionCachePath> location of the regionpicker cache
+	RegionCachePath = filepath.Join(".outreach", ".cache", "box")
+	// RegionCacheFile is the name of the regionpicker cache file
+	RegionCacheFile = "regions.json"
+)
+
+// cache is the package specific cache instance used. This is a global variable to
+// ensure that this is go-routine safe.
+var cache = &cacheStore{}
+
+// cacheEntry is information on a cloud's cache entry
+type cacheEntry struct {
+	// Duration is how long it took to talk to this region at LastUpdatedAt
+	Duration time.Duration `json:"duration"`
+
+	// ExpiresAt is when this cache entry expires. UTC.
+	ExpiresAt time.Time `json:"expires_at"`
+}
+
+type cacheStore struct {
+	// Clouds is a map of cloudName to cache entries
+	// cloud -> region -> cacheEntry
+	Clouds map[CloudName]map[Name]cacheEntry `json:"clouds"`
+
+	mu sync.Mutex
+}
+
+// Get returns the duration of a cloud/region pairing, if it exists.
+// Otherwise ok is returned as false and time.Duration is it's zero-value.
+func (c *cacheStore) Get(cloud CloudName, r Name) (time.Duration, bool) {
+	if c.Clouds == nil {
+		c.load()
+	}
```

Comment:
you may just want to use a sync.Once for this and the other places you did this

Reply:
Yeah, this is not concurrency safe.  You can't even inspect `c.Clouds` without a lock.   I suggest simplifying the whole thing by following these rules:

1. Every exported function just does `c.mu.Lock(); defer c.mu.Unlock(); c.init()` at the top.
2. All unexported functions know the lock is held and can do whatever
3. define `init()` to initialize c.Clouds if it is not already defined.

Reply:
You could also use a `sync.RWMutex` or a `sync.Map` to try and squeeze every little bit of perf out of it if you really cared

Reply:
good catch, I meant to guard against this